sa/storage-authority.go

```diff
@@ -229,6 +230,73 @@ func (ssa *SQLStorageAuthority) GetLatestValidAuthorization(registrationID int64
 	return ssa.GetAuthorization(auth.ID)
 }
 
+// incrementIP increments the byte within `ip` at `index`, accounting for overflow.
+// Index must be greater than zero and less than len(ip).
+func incrementIP(ip *net.IP, index int) {
+	if (*ip)[index] == 255 {
+		(*ip)[index-1]++
+		(*ip)[index] = 0
+	} else {
+		(*ip)[index]++
+	}
+}
+
+// ipRange returns a range of IP addresses suitable for querying MySQL for the
+// purpose of rate limiting using a range that is inclusive on the lower end and
+// exclusive at the higher end. If ip is an IPv4 address, it returns that address,
+// plus the one immediately higher than it. If ip is an IPv6 address, it applies
+// a /48 mask to it and returns the lowest IP in the resulting network, and the
+// first IP outside of the resulting network.
+func ipRange(ip net.IP) (net.IP, net.IP) {
+	ip = ip.To16()
+	begin := make(net.IP, len(ip))
+	end := make(net.IP, len(ip))
+	// For IPv4 addresses, do a match on exact address, so begin = ip and end =
+	// next higher IP.
+	if ip.To4() != nil {
+		copy(begin, ip)
+		copy(end, ip)
+		incrementIP(&end, 15)
+	} else {
+		// For IPv6, match on a certain subnet range, since one person can commonly
+		// have an entire /48 to themselves.
+		maskLength := 48
+		mask := net.CIDRMask(maskLength, 128)
+		begin = ip.Mask(mask)
+		copy(end, begin)
+		incrementIP(&end, (maskLength/8)-1)
+	}
+	return begin, end
+}
+
+// CountRegistrationsByIP returns the number of registrations created in the
+// time range in an IP range. For IPv4 addresses, that range is limited to the
+// single IP. For IPv6 addresses, that range is a /48, since it's not uncommon
+// for one person to have a /48 to themselves.
+func (ssa *SQLStorageAuthority) CountRegistrationsByIP(ip net.IP, earliest time.Time, latest time.Time) (int, error) {
```

Comment:
Aside: I figured out what you were doing here, and then dug through the MariaDB documentation just *absolutely sure* that by this day and age they'd have added query types for IP/CIDR. 

And then my faith in humanity was damaged, again.